csharp/rust/src/lib.rs

```diff
@@ -1,284 +1,472 @@
+    let client_ptr = unsafe { Box::from_raw(in_client_ptr as *mut FFIHandle) };
+    let _runtime_handle = client_ptr.runtime.enter();
+    drop(client_ptr);
+    logger_core::log(Trace, "csharp_ffi", "Exiting csharp_free_client_handle");
 }
 
-/// Unsafe function because creating string from pointer.
+/// # Summary
+/// Method to invoke a command.
+///
+/// # Params
+/// ***in_client_ptr*** An active client handle
+/// ***in_callback*** A callback method with the signature:
+///                   `void Callback(void * in_data, int out_success, const Value value)`.
+///                   The first arg contains the data of the parameter *in_callback_data*;
+///                   the second arg indicates whether the third parameter contains the error or result;
+///                   the third arg contains either the result and MUST be freed by the callback.
+/// ***in_callback_data*** The data to be passed in to *in_callback*.
+/// ***in_request_type*** The type of command to issue.
+/// ***in_routing_info*** Either nullptr or the routing info to use for the command.
+/// ***in_args*** A C-String array of arguments to be passed, with the size of `in_args_count` and zero terminated.
+/// ***in_args_count*** The number of arguments in *in_args*.
+///
+/// # Input Safety (in_...)
+/// The data passed in is considered "caller responsibility".
+/// Any pointers hence will be left unreleased after leaving.
 ///
-/// # Safety
+/// # Output Safety (out_... / return ...)
+/// The data returned is considered "caller responsibility".
+/// The caller must release any non-null pointers.
 ///
-/// * `file_name` must not be `null`.
-/// * `file_name` must be able to be safely casted to a valid [`CStr`] via [`CStr::from_ptr`]. See the safety documentation of [`std::ffi::CStr::from_ptr`].
+/// # Reference Safety (ref_...)
+/// Any reference data is considered "caller owned".
+///
+/// # Freeing data allocated by the API
+/// To free data returned by the API, use the corresponding `free_...` methods of the API.
+/// It is **not optional** to call them to free data allocated by the API!
 #[no_mangle]
-#[allow(improper_ctypes_definitions)]
-pub unsafe extern "C" fn init(level: Option<Level>, file_name: *const c_char) -> Level {
-    let file_name_as_str;
-    unsafe {
-        file_name_as_str = if file_name.is_null() {
-            None
-        } else {
-            Some(
-                CStr::from_ptr(file_name)
-                    .to_str()
-                    .expect("Can not read string argument."),
-            )
+pub extern "C-unwind" fn csharp_command(
+    in_client_ptr: *const c_void,
+    in_callback: CommandCallback,
+    in_callback_data: *mut c_void,
+    in_request_type: RequestType,
+    in_routing_info: *const routing::RoutingInfo,
+    // ToDo: Rework into parameter struct (understand how Command.arg(...) works first)
+    //       handling the different input types.
+    in_args: *const *const c_char,
+    in_args_count: c_int,
```

Comment:
This doesn't support binary strings. Please copy from `main` (there are 3 args).

Reply:
Forgot to integrate them apparently. It was on my list. Will fix.